ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl

```diff
@@ -0,0 +1,124 @@
+### Homotopy Continuation Based Steady State Finding ###
+
+"""
+    hc_steady_states(rs::ReactionSystem, ps; filter_negative=true, neg_thres=-1e-20, u0=typeof(ps)(), kwargs...)
+
+Uses homotopy continuation to find the steady states of the ODE system corresponding to the provided reaction system.
+
+Arguments:
+- `rs::ReactionSystem`: The reaction system for which we want to find the steady states.
+- `ps`: The parameter values for which we want to find the steady states.
+- `filter_negative=true`: If set to true, solutions with any species concentration <0 is removed from the output.
+- `neg_thres=-1e-20`: Determine the minimum values for which a species concentration is to be considred non-negative. Species conentrations ``> neg_thres`` but `< 0.0` are set to `0.0`.
+- `u0=typeof(ps)()`: Initial conditions for which we want to find the steady states. For systems with conservation laws this are required to compute conserved quantities.
+- `kwargs...`: any additional arguments (like `show_progress= true`) are passed into HomotopyContinuation.jl's `solve` call. 
+
+Examples:
+```@repl
+rs = @reaction_network begin
+    k1, Y --> 2X
+    k2, 2X --> X + Y
+    k3, X + Y --> Y
+    k4, X --> 0
+end
+ps = [:k3 => 1.0, :k2 => 2.0, :k4 => 1.5, :k1=>8.0]
+hc_sol = hc_steady_states(rs, ps)
+```
+gives
+```
+[0.5000000000000002, 2.0000000000000004]
+[0.0, 0.0]
+[4.499999999999999, 5.999999999999999]
+```
+
+Notes:
+- This is a wrapper around the `solve` function provided by HomotopyContinuation.jl, all credit for this functionality to that package's authors.
+```
+  """
+function Catalyst.hc_steady_states(rs::ReactionSystem, ps; filter_negative=true, neg_thres=-1e-20, u0=typeof(ps)(), kwargs...)
+    ss_poly = steady_state_polynomial(rs, ps, u0)
+    sols = HC.real_solutions(HC.solve(ss_poly; kwargs...))
+    reorder_sols!(sols, ss_poly, rs)
+    return (filter_negative ? filter_negative_f(sols; neg_thres=neg_thres) : sols)
+end
+
+# For a given reaction system, paraemter values, and initial conditions, find the polynomial that HC solves to find steady states.
+function steady_state_polynomial(rs::ReactionSystem, ps, u0)
+    ns = convert(NonlinearSystem, rs; remove_conserved = true)
+    pre_varmap = map(pair -> pair::Pair{Num,Float64}, [symmap_to_varmap(rs,u0); symmap_to_varmap(rs,ps)]) 
+    conservationlaw_errorcheck(rs, pre_varmap)
+    p_vals = ModelingToolkit.varmap_to_vars(pre_varmap, parameters(ns); defaults = ModelingToolkit.defaults(ns))
+    p_dict  = Dict(parameters(ns) .=> p_vals)
+    eqs_pars_funcs = vcat(equations(ns), conservedequations(rs))
+    eqs_funcs = map(eq -> substitute(eq.rhs - eq.lhs, p_dict), eqs_pars_funcs)
+    eqs = [deregister([mm, mmr, hill, hillr, hillar], eq) for eq in eqs_funcs]
```

Comment:
I've kind of wondered for a while if we should just not register all these and let them be substituted in directly by Symbolics? i.e. if one does 
```julia
mm2(X,v,K) = v*X/(X+K)
rn2 = @reaction_network a begin
       mm2(X,v,K), A --> B
       end
equations(rn2)
```
one gets
```julia
julia> equations(rn2)
1-element Vector{Union{Equation, Reaction}}:
 (X*v) / (K + X), A --> B
```
I guess the disadvantage is we can't autodetect them if we want to know about their presence, but we don't do that currently anyways.

Reply:
wasn't there some advantage that we have recorded the various derivatives of `mm` etc. , so these are known? Maybe the yare calculated at instant speed anyway though. I have no strong opinions, and maybe we should open an issue to discuss the pros/cons (I will mostly defer to your opinions though)

Reply:
I can't remember why we did this change either unfortunately.

Reply:
The only thought I have is we can programmatically check for hill equations via the function type, which Chris and I discussed adding a JumpProcesses specialization for.